default/AI/ResearchAI.py

```diff
@@ -20,19 +22,396 @@ def get_research_index():
+        if not this_spec:
+            continue
+        this_stealth = stealth_grades.get(get_ai_tag_grade(list(this_spec.tags), "STEALTH"), 0)
+        if this_stealth > stealth:
+            stealth_species = specName
+            stealth = this_stealth
+    result = (stealth_species, stealth)
+    return result
+
+def get_ship_tech_usefulness(tech, ship_designer):
+    this_tech = fo.getTech(tech)
+    if not this_tech:
+        print "Invalid Tech specified"
+        return 0
+    unlocked_items = this_tech.unlockedItems
+    unlocked_hulls = []
+    unlocked_parts = []
+    for item in unlocked_items:
+        if item.type == fo.unlockableItemType.shipPart:
+            unlocked_parts.append(item.name)
+        elif item.type == fo.unlockableItemType.shipHull:
+            unlocked_hulls.append(item.name)
+    if not (unlocked_parts or unlocked_hulls):
+        return 0
+    old_designs = ship_designer.optimize_design(consider_fleet_count=False)
+    new_designs = ship_designer.optimize_design(additional_hulls=unlocked_hulls, additional_parts=unlocked_parts, consider_fleet_count=False)
+    if not (old_designs and new_designs):
+        # AI is likely defeated; don't bother with logging error message
+        return 0
+    old_rating, old_pid, old_design_id, old_cost = old_designs[0]
+    old_design = fo.getShipDesign(old_design_id)
+    old_rating = old_rating
+    new_rating, new_pid, new_design_id, new_cost = new_designs[0]
+    new_design = fo.getShipDesign(new_design_id)
+    new_rating = new_rating
+    if new_rating > old_rating:
+        ratio = (new_rating - old_rating) / (new_rating + old_rating)
+        return ratio * 1.5 + 0.5
+    else:
+        return 0
+
+def get_defense_priority(rng):
+    if foAI.foAIstate.aggression <= fo.aggression.cautious:
+        print "AI is cautious. Increasing priority for defense techs."
+        return 2
+    if foAI.foAIstate.misc.get('enemies_sighted', {}):
+        return 1
+    else:
+        return 0.2
+
+def get_production_boost_priority(rng):
+    return 1.5
+
+def get_research_boost_priority(rng):
+    return 2
+
+def get_production_and_research_boost_priority(rng):
+    return 2.5
+
+def get_population_boost_priority(rng):
+    return 2
+
+def get_supply_boost_priority(rng):
+    # TODO consider starlane density and planet density
+    return 1
+
+def get_meter_change_boost_priority(rng):
+    return 1
+
+def get_detection_priority(rng):
+    # TODO consider stealth of enemies
+    return 1
+
+def get_weapon_priority(rng):
+    if foAI.foAIstate.misc.get('enemies_sighted', {}):
+        return 1
+    else:
+        return 0.1
+
+def get_armor_priority(rng):
+    if foAI.foAIstate.misc.get('enemies_sighted', {}):
+        return 1
+    else:
+        return 0.1
+
+def get_shield_priority(rng):
+    if foAI.foAIstate.misc.get('enemies_sighted', {}):
+        return 1
+    else:
+        return 0.1
+
+def get_engine_priority(rng):
+    return 1 if rng.random() < 0.7 else 0
```

Comment:
It is intended for different techs to use different numbers, but not per turn. So lets roll everything right after seeding the RNG then...